File: exercise-4/src/lu.rs
use std::mem::swap;

use crate::scalar::*;

fn dot_product(x: &[Scalar], y: &[Scalar]) -> Scalar {
    debug_assert!(x.len() == y.len());

    let mut r = SCALAR_ZERO;

    for i in 0..x.len() {
        r += x[i] * y[i];
    }

    r
}

fn triangular_solve<const UPPER: bool>(
    matrix: &Vec<Scalar>,
    permutation: &Vec<Index>,
    x: &mut Vec<Scalar>,
) {
    let dimension = permutation.len();

    for kk in 0..dimension {
        let k = if UPPER {
            permutation[dimension - 1 - kk]
        } else {
            permutation[kk]
        };

        let v = x[k];
        x[k] = SCALAR_ZERO;
        x[k] = (v - dot_product(&matrix[k * dimension..k * dimension + dimension], &x))
            / matrix[k * dimension + k];
    }
}

#[derive(Debug, Clone)]
pub struct LUDecomposition {
    // matricies are in row-major format
    lower_: Vec<Scalar>,
    upper_: Vec<Scalar>,

    permutation_: Vec<Index>,
    inverse_permutation_: Vec<Index>,
}

impl LUDecomposition {
    fn new_zero(dimension: Index) -> LUDecomposition {
        let lower = vec![SCALAR_ZERO; dimension * dimension];
        let upper = vec![SCALAR_ZERO; dimension * dimension];

        let permutation = vec![0; dimension];
        let inverse_permutation = vec![0; dimension];

        LUDecomposition {
            lower_: lower,
            upper_: upper,
            permutation_: permutation,
            inverse_permutation_: inverse_permutation,
        }
    }

    // input matrix is in row-major format
    pub fn compute(matrix: &Vec<Scalar>, dimension: Index) -> LUDecomposition {
        let mut decomposition = LUDecomposition::new_zero(dimension);
        let mut row_used = vec![false; dimension];

        // setting permutations to identity
        for k in 0..dimension {
            decomposition.permutation_[k] = k;
            decomposition.inverse_permutation_[k] = k;
        }

        // setting L to identity
        for k in 0..dimension {
            decomposition.lower_[dimension * k + k] = SCALAR_ONE;
        }

        // computing LU
        for k in 0..dimension {
            // computing the column
            let mut column = vec![SCALAR_ZERO; dimension];

            for i in 0..dimension {
                column[i] = matrix[dimension * i + k];
            }

            triangular_solve::<false>(
                &decomposition.lower_,
                &decomposition.permutation_,
                &mut column,
            );

            // choosing pivot
            let pivot_row = {
                let mut pivot_row = 0;

                for i in 0..dimension {
                    if !row_used[i]
                        && (row_used[pivot_row] || column[i].abs() > column[pivot_row].abs())
                    {
                        pivot_row = i;
                    }
                }

                pivot_row
            };
            row_used[pivot_row] = true;
            let pivot_value = column[pivot_row];

            // computing L and U
            for i in 0..dimension {
                if row_used[i] {
                    decomposition.upper_[dimension * i + pivot_row] = column[i];
                } else {
                    decomposition.lower_[dimension * i + pivot_row] = column[i] / pivot_value;
                }
            }

            // computing the permutation
            let ipr = decomposition.inverse_permutation_[pivot_row];
            let pk = decomposition.permutation_[k];

            decomposition.permutation_[k] = pivot_row;
            decomposition.inverse_permutation_[pivot_row] = k;
            decomposition.permutation_[ipr] = pk;
            decomposition.inverse_permutation_[pk] = ipr;
        }

        decomposition
    }

    pub fn solve(&self, x: &mut Vec<Scalar>) {
        triangular_solve::<false>(&self.lower_, &self.permutation_, x);
        triangular_solve::<true>(&self.upper_, &self.permutation_, x);

        let mut temp = vec![SCALAR_ZERO; x.len()];

        for i in 0..x.len() {
            temp[i] = x[self.permutation_[i]];
        }

        swap(&mut temp, x);
    }
}

File: exercise-6/src/lu.zig
const config = @import("config.zig");
const matrix = @import("matrix.zig");
const utils = @import("utils.zig");
const std = @import("std");

const Scalar = config.Scalar;
const Index = config.Index;

pub const LUDecomposition = struct {
    const Self = @This();

    const MatrixStorage = matrix.RowMajorMatrix;
    const WorkVector = matrix.Vector;

    lower_: MatrixStorage,
    upper_: MatrixStorage,
    permutation_: matrix.Permutation,

    work_row_used_: []bool,
    work_vector_: WorkVector,

    pub fn init(allocator: std.mem.Allocator, dimension: Index) !Self {
        return Self{
            .lower_ = try MatrixStorage.init(allocator, dimension),
            .upper_ = try MatrixStorage.init(allocator, dimension),
            .permutation_ = try matrix.Permutation.init(allocator, dimension),
            .work_row_used_ = try allocator.alloc(bool, dimension),
            .work_vector_ = try matrix.Vector.init(allocator, dimension),
        };
    }

    pub fn deinit(self: *Self, allocator: std.mem.Allocator) void {
        self.lower_.deinit(allocator);
        self.upper_.deinit(allocator);
        self.permutation_.deinit(allocator);

        allocator.free(self.work_row_used_);
        self.work_vector_.deinit(allocator);
    }

    // true if computes, false otherwise
    pub fn factorize(
        self: *Self,
        comptime Matrix: type,
        base: Matrix,
    ) bool {
        const dimension = self.lower_.dimension();

        // memory stuff
        var row_used = self.work_row_used_;
        @memset(row_used[0..dimension], false);

        var work_vector = self.work_vector_;
        work_vector.set_zero();

        self.permutation_.set_identity();
        self.lower_.set_zero();
        self.upper_.set_zero();

        // setting L to identity
        for (0..dimension) |i| {
            self.lower_.at_mut(i, i).* = 1;
        }

        // computing LU
        for (0..dimension) |col| {
            for (0..dimension) |i| {
                work_vector.at_mut(i).* = base.at(i, col);
            }

            utils.triangular_solve(
                MatrixStorage,
                matrix.Vector,
                utils.Triangularity.Lower,
                self.lower_,
                self.permutation_,
                &work_vector,
            );

            var pivot_row: Index = 0;

            for (0..dimension) |i| {
                if (!row_used[i] and
                    (row_used[pivot_row] or
                    @abs(work_vector.at(i)) > @abs(work_vector.at(pivot_row))))
                {
                    pivot_row = i;
                }
            }

            const pivot_value = work_vector.at(pivot_row);

            if (@abs(pivot_value) < config.MIN_PIVOT) {
                return false;
            }

            row_used[pivot_row] = true;

            for (0..dimension) |i| {
                if (row_used[i]) {
                    self.upper_.at_mut(i, pivot_row).* =
                        work_vector.at(i);
                } else {
                    self.lower_.at_mut(i, pivot_row).* =
                        work_vector.at(i) / pivot_value;
                }
            }

            self.permutation_.consistent_set(col, pivot_row);
        }

        return true;
    }

    pub fn solve(
        self: *Self,
        comptime VectorType: type,
        vector: *VectorType,
    ) void {
        utils.triangular_solve(
            MatrixStorage,
            matrix.Vector,
            utils.Triangularity.Lower,
            self.lower_,
            self.permutation_,
            vector,
        );

        utils.triangular_solve(
            MatrixStorage,
            matrix.Vector,
            utils.Triangularity.Upper,
            self.upper_,
            self.permutation_,
            vector,
        );

        const dimension = self.lower_.dimension();

        for (0..dimension) |i| {
            self.work_vector_.at_mut(i).* = vector.at(self.permutation_.perm(i));
        }

        for (0..dimension) |i| {
            vector.at_mut(i).* = self.work_vector_.at(i);
        }
    }
};
